### PR TITLE
Refactor cli_test #1192

### PR DESCRIPTION
**Pull request by cogutvalera**
PR for #1192

### CODE REVIEW

File: tests/cli/main.cpp [hunk 1]
```diff
@@ -196,253 +196,190 @@ class client_connection
...
 ///////////////////
 // Test blind transactions and mantissa length of range proofs.
 ///////////////////
-BOOST_AUTO_TEST_CASE( cli_confidential_tx_test )
+BOOST_FIXTURE_TEST_CASE( cli_confidential_tx_test, cli_fixture )
 {
-   using namespace graphene::chain;
-   using namespace graphene::app;
+   //using namespace graphene::chain;
+   //using namespace graphene::app;
```
Comment (oxarbitrage) [inline tests/cli/main.cpp]: pls remove comments.

**Comment by cogutvalera** [comment]
Thanks !

File: tests/cli/main.cpp (hunk 1 above)
Comment (cogutvalera) [reply tests/cli/main.cpp]: sure ! removed ! thanks !

**Comment by cogutvalera** [comment]
can we merge this PR and close related issue ? or this is not still ready and need some changes and improvements ? if YES, then which changes and improvements in your opinion ?

Thanks !

**Comment by abitmore** [comment]
I'd say split it into smaller files.

File: tests/cli/main.cpp [hunk 2]
```diff
@@ -196,253 +196,188 @@ class client_connection
...
-
-      // wait for everything to finish up
-      fc::usleep(fc::seconds(1));
-   } catch( fc::exception& e ) {
-      edump((e.to_detail_string()));
-      throw;
-   }
-   app1->shutdown();
+    try
+    {
```
Comment (abitmore) [inline tests/cli/main.cpp]: Why changed the indentation to 4 white spaces (both above and below)?

Comment (cogutvalera) [reply tests/cli/main.cpp]: where can I see BitShares Core coding convention ? I prefer 4 white spaces, what do you prefer ?

File: tests/cli/main.cpp [hunk 3]
```diff
@@ -196,253 +196,188 @@ class client_connection
...
+    client_connection* con = new client_connection(app1, app_dir, server_port_number);
+    std::vector<std::string> nathan_keys { "5KQwrPbwdL6PhXujxW37FSSQZ1JiwsST4cqQzDeyXtP79zkvFD3" };
+
+    std::vector<signed_transaction> import_txs;
+    account_object nathan_acct_before_upgrade;
+    signed_transaction upgrade_tx;
+    account_object nathan_acct_after_upgrade;
+    graphene::wallet::brain_key_info bki;
+    signed_transaction create_acct_tx;
+    signed_transaction transfer_tx;
```
Comment (abitmore) [inline tests/cli/main.cpp]: Should not define so many different transaction objects in `cli_fixture`; `nathan_acct_before_upgrade` and `nathan_acct_after_upgrade` should not be here as well; they should be declared in individual test cases.

File: tests/cli/main.cpp [hunk 4]
```diff
@@ -196,253 +196,188 @@ class client_connection
...
+            con->wallet_api_ptr->unlock("supersecret");
+
+            // import Nathan account
+            BOOST_TEST_MESSAGE("Importing nathan key");
+            BOOST_CHECK_EQUAL(nathan_keys[0], "5KQwrPbwdL6PhXujxW37FSSQZ1JiwsST4cqQzDeyXtP79zkvFD3");
+            BOOST_CHECK(con->wallet_api_ptr->import_key("nathan", nathan_keys[0]));
+
+            BOOST_TEST_MESSAGE("Importing nathan's balance");
+            import_txs = con->wallet_api_ptr->import_balance("nathan", nathan_keys, true);
+            nathan_acct_before_upgrade = con->wallet_api_ptr->get_account("nathan");
```
Comment (abitmore) [inline tests/cli/main.cpp]: The nathan account upgrading should be in a "basic" test case and be invoked by needed test cases, but not in fixture.

File: tests/cli/main.cpp [hunk 5]
```diff
@@ -196,253 +196,188 @@ class client_connection
...
+            nathan_acct_after_upgrade = con->wallet_api_ptr->get_account("nathan");
+
+            // verify that the upgrade was successful
+            BOOST_CHECK_PREDICATE( std::not_equal_to<uint32_t>(), (nathan_acct_before_upgrade.membership_expiration_date.sec_since_epoch())(nathan_acct_after_upgrade.membership_expiration_date.sec_since_epoch()) );
+            BOOST_CHECK(nathan_acct_after_upgrade.is_lifetime_member());
+
+            // create a new account
+            bki = con->wallet_api_ptr->suggest_brain_key();
+            BOOST_CHECK(!bki.brain_priv_key.empty());
+            create_acct_tx = con->wallet_api_ptr->create_account_with_brain_key(bki.brain_priv_key, "jmjatlanta", "nathan", "nathan", true);
```
Comment (abitmore) [inline tests/cli/main.cpp]: The jmj account creation should be in a test case as well.

File: tests/cli/main.cpp [hunk 6]
```diff
@@ -196,253 +196,188 @@ class client_connection
...
+            // save the private key for this new account in the wallet file
+            BOOST_CHECK(con->wallet_api_ptr->import_key("jmjatlanta", bki.wif_priv_key));
+            con->wallet_api_ptr->save_wallet_file(con->wallet_filename);
+
+            // attempt to give jmjatlanta some bitsahres
+            BOOST_TEST_MESSAGE("Transferring bitshares from Nathan to jmjatlanta");
+            transfer_tx = con->wallet_api_ptr->transfer("nathan", "jmjatlanta", "10000", "1.3.0", "Here are some CORE token for your new account", true);
```
Comment (abitmore) [inline tests/cli/main.cpp]: The transfer should not be in cli_fixture.

File: tests/cli/main.cpp (hunk 2 above)
Comment (oxarbitrage) [reply tests/cli/main.cpp]: 3 spaces. sorry @abitmore i didnt saw this one.

Comment (cogutvalera) [reply tests/cli/main.cpp]: indentation everywhere in BitShares Core code base is prefer to 3 spaces ?

Comment (oxarbitrage) [reply tests/cli/main.cpp]: yes. 3 spaces everywhere(except p2p code that uses 2 but we don't want to change that now)

Comment (cogutvalera) [reply tests/cli/main.cpp]: ok sure. Thank you very much ! Does BitShares Core have code convention ?

**Comment by cogutvalera** [comment]
made improvements as @abitmore mentioned, do we need here some more changes/improvements or not ?

File: tests/cli/main.cpp (hunk 2 above)
Comment (abitmore) [reply tests/cli/main.cpp]: Can you please change the indentation back to 3 white spaces? Thanks.

We don't have code conventions so far.

File: tests/cli/main.cpp [hunk 7]
```diff
@@ -197,252 +197,218 @@ class client_connection
...
 {
-   using namespace graphene::chain;
-   using namespace graphene::app;
-   std::shared_ptr<graphene::app::application> app1;
-   try {
-      fc::temp_directory app_dir ( graphene::utilities::temp_directory_path() );
+   int server_port_number {0};
+   fc::temp_directory app_dir = graphene::utilities::temp_directory_path();
+   std::shared_ptr<graphene::app::application> app1 = start_application(app_dir, server_port_number);
+   client_connection* con = new client_connection(app1, app_dir, server_port_number);
```
Comment (abitmore) [inline tests/cli/main.cpp]: Initializing these variables here is a bad idea. They're heavy. Better move them to a constructor. Then some member variables can be changed to local variable.

By the way, why need to use a pointer?

File: tests/cli/main.cpp [hunk 8]
```diff
@@ -197,252 +197,218 @@ class client_connection
...
+        signed_transaction upgrade_tx, transfer_tx;
+
+        BOOST_TEST_MESSAGE("Importing nathan's balance");
+        import_txs = con->wallet_api_ptr->import_balance("nathan", nathan_keys, true);
+        nathan_acct_before_upgrade = con->wallet_api_ptr->get_account("nathan");
+
+        // upgrade nathan
+        BOOST_TEST_MESSAGE("Upgrading Nathan to LTM");
+        upgrade_tx = con->wallet_api_ptr->upgrade_account("nathan", true);
+        nathan_acct_after_upgrade = con->wallet_api_ptr->get_account("nathan");
```
Comment (abitmore) [inline tests/cli/main.cpp]: The upgrade of nathan account can be abstracted to a single test case, then be invoked by other test cases. See this for example: https://github.com/bitshares/bitshares-core/blob/610dbf519aa2a97523a6d753df93a00f368e4f8e/tests/tests/operation_tests.cpp#L983

File: tests/cli/main.cpp (hunk 7 above)
Comment (cogutvalera) [reply tests/cli/main.cpp]: 1. fixed indentation from 4 white spaces to 3
2. initializing variables in constructor has different compile errors, need to fight all them, trying to understand why they happened, still fighting ...
3. I've used pointer because without pointer I had compile errors, still trying to understand why and still fighting with them ...

File: tests/cli/main.cpp (hunk 8 above)
Comment (cogutvalera) [reply tests/cli/main.cpp]: will do it soon

Comment (cogutvalera) [reply tests/cli/main.cpp]: INVOKE doesn't work well enough with test fixtures, also need to fight here with errors

File: tests/cli/main.cpp (hunk 7 above)
Comment (cogutvalera) [reply tests/cli/main.cpp]: unknown location(0): fatal error: in "cli_connect": memory access violation at address: 0x00000064: no mapping at fault address
/home/valera/Projects/Blockchain/Graphene/BitShares/bitshares-core/tests/cli/main.cpp(260): last checkpoint: "cli_connect" fixture ctor

*** 1 failure is detected in the test module "Test Application"

such error I see when trying to call next method:

`app_dir = graphene::utilities::temp_directory_path();`

It means next:

`BOOST_FIXTURE_TEST_CASE( cli_connect, cli_fixture )` - this Boost fixture cannot call `cli_fixture` constructor with `graphene::utilities::temp_directory_path()` method inside, maybe
I'm doing something wrong ? Still trying to understand the reason and to code as we desire !

Comment (cogutvalera) [reply tests/cli/main.cpp]: fixed by direct initialization instead of using assignment operators inside constructor body

**Comment by cogutvalera** [comment]
guys when you will have enough time check please my new changes, I've committed all required changes that were asked by @abitmore , hope all was done correctly enough, I've compiled and tested and it seems work absolutely as intended.

Thanks a lot !

File: tests/cli/main.cpp [hunk 9]
```diff
@@ -505,65 +474,25 @@ BOOST_AUTO_TEST_CASE( cli_confidential_tx_test )
...
-                             (nathan_acct_after_upgrade.membership_expiration_date.sec_since_epoch()) );
-      BOOST_CHECK(nathan_acct_after_upgrade.is_lifetime_member());
+      INVOKE(upgrade_nathan_account);
 
       // create a new account
       graphene::wallet::brain_key_info bki = con.wallet_api_ptr->suggest_brain_key();
       BOOST_CHECK(!bki.brain_priv_key.empty());
-      signed_transaction create_acct_tx = con.wallet_api_ptr->create_account_with_brain_key(
-                                                bki.brain_priv_key, "jmjatlanta", "nathan", "nathan", true);
+      signed_transaction create_acct_tx = con.wallet_api_ptr->create_account_with_brain_key(bki.brain_priv_key, "jmjatlanta", "nathan", "nathan", true);
```
Comment (abitmore) [inline tests/cli/main.cpp]: Apparently your screen is wider than mine. I like the wrapped format, at max around 130 characters a line.

File: tests/cli/main.cpp [hunk 10]
```diff
@@ -197,102 +199,118 @@ class client_connection
...
+      nathan_acct_before_upgrade = con.wallet_api_ptr->get_account("nathan");
 
-      BOOST_CHECK_THROW( con.wallet_api_ptr->quit(), fc::canceled_exception );
+      // upgrade nathan
+      BOOST_TEST_MESSAGE("Upgrading Nathan to LTM");
+      upgrade_tx = con.wallet_api_ptr->upgrade_account("nathan", true);
+      nathan_acct_after_upgrade = con.wallet_api_ptr->get_account("nathan");
+
+      // verify that the upgrade was successful
+      BOOST_CHECK_PREDICATE( std::not_equal_to<uint32_t>(), (nathan_acct_before_upgrade.membership_expiration_date.sec_since_epoch())(nathan_acct_after_upgrade.membership_expiration_date.sec_since_epoch()) );
```
Comment (abitmore) [inline tests/cli/main.cpp]: This line is too long to fit in my screen as well.

By the way there were some long lines that didn't change, so I'm unable to comment on them. Please help wrap them as well.

File: tests/cli/main.cpp [hunk 11]
```diff
@@ -197,102 +199,118 @@ class client_connection
...
+   int server_port_number;
+   fc::temp_directory app_dir;
    std::shared_ptr<graphene::app::application> app1;
-   try {
-      fc::temp_directory app_dir ( graphene::utilities::temp_directory_path() );
+   client_connection con;
+   std::vector<std::string> nathan_keys;
+
+   cli_fixture() : 
+   server_port_number(0),
```
Comment (abitmore) [inline tests/cli/main.cpp]: I think we can use some indentations here.

File: tests/cli/main.cpp (hunk 9 above)
Comment (cogutvalera) [reply tests/cli/main.cpp]: ok sure ! I will format the whole file right now as required !

File: tests/cli/main.cpp (hunk 10 above)
Comment (cogutvalera) [reply tests/cli/main.cpp]: ok sure ! I will format the whole file right now as required !

File: tests/cli/main.cpp (hunk 11 above)
Comment (cogutvalera) [reply tests/cli/main.cpp]: ok sure !

**Comment by cogutvalera** [comment]
Thank you !

**Comment by oxarbitrage** [comment]
i think it is looking great but i will suggest the following if possible:

the "create a new account", "save the private key for this new account in the wallet file" and "attempt to give jmjatlanta some bitsahres" are pretty common actions. maybe create test `createaccount` and add the 3 things there. then we can invoke this new test after `upgrade_nathan_account`.

for the new test to make sense you can add some checks on it like if the account was created and if the transfer went ok.

**Comment by cogutvalera** [comment]
@oxarbitrage thank you very much ! Your idea great and make sense ! I think I need to code it too. Is anybody disagree ?

File: tests/cli/main.cpp [hunk 12]
```diff
@@ -506,21 +501,11 @@ BOOST_FIXTURE_TEST_CASE( account_history_pagination, cli_fixture )
...
-      );
-      // save the private key for this new account in the wallet file
-      BOOST_CHECK(con.wallet_api_ptr->import_key("jmjatlanta", bki.wif_priv_key));
-      con.wallet_api_ptr->save_wallet_file(con.wallet_filename);
+      INVOKE(create_new_account);
 
       // attempt to give jmjatlanta some bitsahres
       BOOST_TEST_MESSAGE("Transferring bitshares from Nathan to jmjatlanta");
-      for(int i = 1; i <= 200; i++)
+      for(int i = 1; i <= 199; i++)
```
Comment (abitmore) [inline tests/cli/main.cpp]: Why this change? Something was broken?

Comment (cogutvalera) [reply tests/cli/main.cpp]: nothing was broken, I've changed this because we have +1 transaction inside INVOKE create_new_account, and I've decided to decrease here history by -1 transaction. Should I change another way ? I can change BOOST_CHECK history transfers to 202 as an example here
https://github.com/bitshares/bitshares-core/blob/18a78a5e614c941dbdafd3219b6476efdac9c13a/tests/cli/main.cpp#L518

what are your thoughts ? Thanks !

Comment (abitmore) [reply tests/cli/main.cpp]: >I've changed this because we have +1 transaction inside INVOKE create_new_account, and I've decided to decrease here history by -1 transaction. 

Makes sense. Thanks.

**Comment by pmconrad** [comment]
Looks good, but haven't checked in depth.

**Comment by cogutvalera** [comment]
Thank you guys !

**Comment by cogutvalera** [comment]
Thanks for merge !
